consensus/ibft/bridge.go

```diff
@@ -0,0 +1,173 @@
+}
+
+// updateValidatorsHook updates validator set in Bridge
+func (b *BridgeMechanism) updateValidatorsHook(snapParam interface{}) error {
+	// Cast the param to a *Snapshot
+	snap, ok := snapParam.(*Snapshot)
+	if !ok {
+		return ErrInvalidHookParam
+	}
+
+	b.bridge.SetValidators(snap.Set, b.calculateSignatureThreshold(snap.Set))
+
+	return nil
+}
+
+// verifyStateTransactionsHook verify state transactions in the block
+func (b *BridgeMechanism) verifyStateTransactionsHook(blockParam interface{}) error {
+	block, ok := blockParam.(*types.Block)
+	if !ok {
+		return ErrInvalidHookParam
+	}
+
+	for _, tx := range block.Transactions {
+		if tx.Type != types.TxTypeState {
+			continue
+		}
+
+		if err := b.bridge.ValidateTx(tx); err != nil {
+			b.ibft.logger.Error("block verification failed, block has invalid state transactions", "err", err)
+
+			return errBlockVerificationFailed
+		}
+	}
+
+	return nil
+}
+
+// insertTransactionHookParams are the params passed into the InsertTransactionsHook
+type insertTransactionHookParams struct {
+	header       *types.Header
+	transition   *state.Transition
+	transactions *[]*types.Transaction
+}
+
+// insertStateTransactionsHook applyes state transactions and adds to block
+func (b *BridgeMechanism) insertStateTransactionsHook(rawParams interface{}) error {
+	params, ok := rawParams.(*insertTransactionHookParams)
+	if !ok {
+		return ErrInvalidHookParam
+	}
+
+	msgs, err := b.bridge.GetReadyMessages()
+	if err != nil {
+		b.ibft.logger.Warn("failed to get ready messages from bridge", "err", err)
+
+		return nil
+	}
+
+	signatureThreshold := b.calculateSignatureThreshold(b.ibft.state.validators)
+	for _, msg := range msgs {
+		if uint64(len(msg.Signatures)) < signatureThreshold {
+			b.ibft.logger.Warn(
+				"message doesn't have enough signatures",
+				"wanted", signatureThreshold,
+				"actual", len(msg.Signatures),
+			)
+
+			continue
+		}
+
+		signer := crypto.NewSigner(
+			b.ibft.config.Params.Forks.At(b.ibft.state.view.Sequence),
+			uint64(b.ibft.config.Params.ChainID),
+		)
+
+		tx, err := signer.SignTx(msg.Transaction, b.ibft.validatorKey)
+		if err != nil {
+			return fmt.Errorf("failed to sign state transaction: %w", err)
+		}
+
+		if err := params.transition.Write(tx); err != nil {
+			b.ibft.logger.Warn("failed to apply state transaction", "tx", tx, "err", err)
+
+			continue
+		}
+
+		*params.transactions = append(*params.transactions, tx)
+	}
+
+	return nil
+}
+
+// consumeStateTransactionsHook consumes all state transactions added in the block
+func (b *BridgeMechanism) consumeStateTransactionsHook(numberParam interface{}) error {
+	block, ok := numberParam.(*types.Block)
+	if !ok {
+		return ErrInvalidHookParam
+	}
+
+	for _, tx := range block.Transactions {
+		if tx.Type != types.TxTypeState {
+			continue
+		}
+
+		b.bridge.Consume(tx)
+	}
+
+	return nil
+}
+
+func (b *BridgeMechanism) calculateSignatureThreshold(set ValidatorSet) uint64 {
+	return uint64(set.Len())
+}
```

Comment:
Why is the signature threshold the entire size of the validator set and not the majority?

This means that every validator needs to sign this message in order for the threshold to be met.

Why not use the provided `NumValid` method instead?

Reply:
Changed at ffcf9323e750de75f601b0e04adc31d5636aaf00